src/client/datascience/data-viewing/data-wrangler/dataWrangler.ts

```diff
@@ -0,0 +1,821 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+'use strict';
+import '../../../common/extensions';
+
+import { inject, injectable, named } from 'inversify';
+import * as path from 'path';
+import * as uuid from 'uuid/v4';
+import { Disposable, EventEmitter, Memento, NotebookCell, ViewColumn, WebviewPanel } from 'vscode';
+
+import {
+    IApplicationShell,
+    ICommandManager,
+    IWebviewPanelProvider,
+    IWorkspaceService,
+    IDocumentManager
+} from '../../../common/application/types';
+import { EXTENSION_ROOT_DIR, PYTHON_LANGUAGE, UseCustomEditorApi } from '../../../common/constants';
+import { traceError } from '../../../common/logger';
+import { GLOBAL_MEMENTO, IConfigurationService, IDisposable, IMemento } from '../../../common/types';
+import * as localize from '../../../common/utils/localize';
+import { Commands, Identifiers } from '../../constants';
+import {
+    ICodeCssGenerator,
+    IJupyterVariableDataProvider,
+    IJupyterVariableDataProviderFactory,
+    IJupyterVariables,
+    IThemeFinder,
+    WebViewViewChangeEventArgs
+} from '../../types';
+import { updateCellCode } from '../../notebook/helpers/executionHelpers';
+import { CssMessages } from '../../messages';
+import { ColumnType, DataViewerMessages, IDataFrameInfo, IDataViewerDataProvider } from '../types';
+import {
+    IDataWrangler,
+    DataWranglerMessages,
+    DataWranglerCommands,
+    IRenameColumnsRequest,
+    IHistoryItem,
+    IDropRequest,
+    INormalizeColumnRequest,
+    IFillNaRequest,
+    IDropDuplicatesRequest,
+    IDropNaRequest,
+    ICoerceColumnRequest,
+    IGetHistoryItem,
+    IReplaceAllColumnsRequest,
+    IRemoveHistoryItemRequest,
+    SidePanelSections,
+    IGetColumnStatsReq,
+    ICellCssStylesHash
+} from './types';
+import { DataScience } from '../../../common/utils/localize';
+import { DataViewer } from '../dataViewer';
+
+const PREFERRED_VIEWGROUP = 'JupyterDataWranglerPreferredViewColumn';
+const dataWranglerDir = path.join(EXTENSION_ROOT_DIR, 'out', 'datascience-ui', 'viewers');
+
+// Keeps track of all the transformations called on the data wrangler
+// Runs the transformations, communicates with the data wrangler UI through onMessage and postMessage
+@injectable()
+export class DataWrangler extends DataViewer implements IDataWrangler, IDisposable {
+    private existingDisposable: Disposable | undefined;
+    private historyList: IHistoryItem[] = [];
+    private sourceFile: string | undefined;
+    private commands = new Map<
+        DataWranglerCommands,
+        // eslint-disable-next-line @typescript-eslint/no-explicit-any
+        (args: any, currentVariableName: string) => Promise<IHistoryItem | void>
```

Comment:
nit: if you want typed arguments you could possibly do something like

```
export type DataWranglerCommandArgs<T extends DataWranglerCommands> = 
  T extends DataWranglerCommands.Describe ?
  IDescribeArgs :
  T extends DataWranglerCommands.Drop ?
  IDropArgs :
  any;

private commands = new Map<
        DataWranglerCommands,
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        <T>(args: DataWranglerCommandArgs<T>, currentVariableName: string) => Promise<IHistoryItem | void>
```

another option would be to add the command to the args interface as well, and have a type that is like
```
export type DataWranglerCommandArgs = IDescribeArgs | IDropArgs;
```